File: 14-LongestCommonPrefix/14-LongestCommonPrefix.py
"""

14. Longest Common Prefix

Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.

Constraints:

    1 <= strs.length <= 200
    0 <= strs[i].length <= 200
    strs[i] consists of only lowercase English letters.

"""

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Solution:
    def insert_in_trie(self, node, word):
        for ch in word:
            if ch not in node.children.keys(): node.children[ch] = TrieNode()
            node = node.children[ch]

    def search(self, node, word):
        res = 0
        for ch in word:
            if ch not in node.children.keys(): return res
            res += 1
            node = node.children[ch]
        return res

    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 1: return strs[0]
        root = TrieNode()
        node = root
        self.insert_in_trie(node, strs[0])
        res = float("inf")
        for i in range(1, len(strs)):
            node = root
            curr = self.search(node, strs[i])
            if curr == 0 : return ""
            res = min(curr, res)
        return strs[0][0:res]

File: 14-LongestCommonPrefix/14-LongestCommonPrefix.java
/*

Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.

Question Link : https://leetcode.com/problems/longest-common-prefix/

*/

class Solution {
    public String longestCommonPrefix(String[] strs) {
        if(strs.length==0)
            return "";
        if (strs.length==1)
            return strs[0].toString();
            
        boolean flag=false;
        StringBuilder sb = new StringBuilder();
        
        Arrays.sort(strs, (a, b)->Integer.compare(a.length(), b.length()));
        
        for(int i=0;i<strs[0].length();i++)
        {
            for(int j=1;j<strs.length;j++)
            {
                if (strs[0].charAt(i)==strs[j].charAt(i))
                    flag=true;
                else
                {
                    flag =false;
                    i=strs[0].length();
                    break;
                }
            }
            if (flag==true)
                    sb.append(strs[0].charAt(i));            
        }

        return sb.toString();
        
    }
}


// Trie Approach

class Solution {
    public String longestCommonPrefix(String[] strs) {
        TrieNode root = new TrieNode();
        TrieNode node = root;
        insertInTrie(strs[0], node);
        int res = strs[0].length();
        for(int i = 1; i < strs.length; i++){
            node = root;
            int curr = search(node, strs[i]);
            if(curr == 0) return "";
            res = Math.min(res, curr);
        }
        return res == 0 ? "" : strs[0].substring(0, res);
    }

    private int search(TrieNode node, String word){
        int res = 0;
        for(char ch : word.toCharArray()){
            if(!node.children.containsKey(ch)) return res;
            res++;
            node = node.children.get(ch);
        }
        return res;
    }

    private void insertInTrie(String word, TrieNode node){
        for(char ch : word.toCharArray()){
            if(!node.children.containsKey(ch)) node.children.put(ch, new TrieNode());
            node = node.children.get(ch);
        }
    }
}

class TrieNode{
    Map<Character, TrieNode> children;
    TrieNode(){
        this.children = new HashMap<>();
    }
}
